server/mqtt.go

```diff
@@ -1338,30 +1372,49 @@ func (s *Server) mqttCreateAccountSessionManager(acc *Account, quitCh chan struc
 			return nil, fmt.Errorf("failed to update stream config: %w", err)
 		}
 	}
-	// Try to transfer regardless if we have already updated the stream or not
-	// in case not all messages were transferred and the server was restarted.
-	if as.transferRetainedToPerKeySubjectStream(s) {
+
+	transferRMS := func() error {
+		if !needToTransfer {
+			return nil
+		}
+
+		as.transferRetainedToPerKeySubjectStream(s)
+
 		// We need another lookup to have up-to-date si.State values in order
 		// to load all retained messages.
 		si, err = lookupStream(mqttRetainedMsgsStreamName, "retained messages")
 		if err != nil {
-			return nil, err
+			return err
 		}
+		needToTransfer = false
+		return nil
+	}
+
+	// Attempt to transfer all "single subject" retained messages to new
+	// subjects. It may fail, will log its own error; ignore it the first time
+	// and proceed to updating MaxMsgsPer. Then we invoke transferRMS() again,
+	// which will get another chance to resolve the error; if not we bail there.
+	if err = transferRMS(); err != nil {
+		return nil, err
 	}
+
 	// Now, if the stream does not have MaxMsgsPer set to 1, and there are no
 	// more messages on the single $MQTT.rmsgs subject, update the stream again.
 	if si.Config.MaxMsgsPer != 1 {
-		_, err := jsa.loadNextMsgFor(mqttRetainedMsgsStreamName, "$MQTT.rmsgs")
-		// Looking for an error indicated that there is no such message.
-		if err != nil && IsNatsErr(err, JSNoMessageFoundErr) {
-			si.Config.MaxMsgsPer = 1
-			// We will need an up-to-date si, so don't use local variable here.
-			if si, err = jsa.updateStream(&si.Config); err != nil {
-				return nil, fmt.Errorf("failed to update stream config: %w", err)
-			}
+		si.Config.MaxMsgsPer = 1
+		// We will need an up-to-date si, so don't use local variable here.
+		if si, err = jsa.updateStream(&si.Config); err != nil {
+			return nil, fmt.Errorf("failed to update stream config: %w", err)
 		}
 	}
 
+	// If we failed the first time, there is now at most one lingering message
+	// in the old subject. Try again (it will be a NO-OP if succeeded the first
+	// time).
+	if err = transferRMS(); err != nil {
```

Comment:
Once the update to maxMsgsPer is 1, what does it mean to call into here now?

Reply:
By now there _may_ be a single message remaining, and we want to delete it. I do not know how to delete it without obtaining the sequence, and that's what the 2nd 1transferRMS` accomplishes. It only errors if it can't get the stream info.